### PR TITLE
Silence warnings on GCC 12

### LINKED ISSUE #2020: warning: ‘index’ may be used uninitialized
Compiling WABT 1.0.30 with gcc 12.2.0, I get the following warning:

```
/usr/include/c++/12/bits/stl_pair.h:665:43: warning: ‘index’ may be used uninitialized [-Wmaybe-uninitialized]
  665 |              || (!(__y.first < __x.first) && __x.second < __y.second); }
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
src/wat-writer.cc: In function ‘wabt::Result src/wat-writer.cc:1717:9: note: ‘index’ was declared here
 1717 |   Index index;
```

### PR DESCRIPTION
**Pull request by keithw**
These changes allow the build to succeed with `-DWERROR=ON` on GCC 12.

Fixes #2020

### CODE REVIEW

File: src/type-checker.cc [hunk 1]
```diff
@@ -172,10 +172,10 @@ Result TypeChecker::GetThisFunctionLabel(Label** label) {
 
 Result TypeChecker::PeekType(Index depth, Type* out_type) {
   Label* label;
+  *out_type = Type::Any;
```
Comment (sbc100) [inline src/type-checker.cc:175]: Why is this needed?

File: src/wat-writer.cc [hunk 2]
```diff
@@ -1743,6 +1743,9 @@ bool WatWriter::IsInlineExport(const Export& export_) {
     case ExternalKind::Tag:
       index = module.GetTagIndex(export_.var);
       break;
+
+    default:
+      WABT_UNREACHABLE;
```
Comment (sbc100) [inline src/wat-writer.cc:1748]: Doesn't adding this default case mean we don't get nice errors when we don't have full case coverage?  (e..g when we add a new extern kind?)

File: src/type-checker.cc (hunk 1 above)
Comment (keithw) [reply src/type-checker.cc:175]: One of the callsites of PeekType (TypeChecker::OnIndexedFuncRef) assumes that it will initialize the `Type*` argument even in case of an `Error` result, but if the `Error` is because of a failure in the `CHECK_RESULT(TopLabel(&label));` line, that won't happen. Moving this line higher makes sure that the `Type*` gets initialized even in that kind of error.

Comment (sbc100) [reply src/type-checker.cc:175]: Isn't the callsite that is wrong then?  I don't think we should need to write to the output variable at all if we are returning an error

File: src/wat-writer.cc (hunk 2 above)
Comment (sbc100) [reply src/wat-writer.cc]: Well, that is shame.  I'd rather not loose those warnings.   Kind of shame that gcc can't figure that out but clang can.

What happens if you add explicit `Fist` and `Last` cases here instead (with a comment explaining why)?

Comment (sbc100) [reply src/wat-writer.cc]: Is this a bug in gcc?

Comment (keithw) [reply src/wat-writer.cc]: > What happens if you add explicit `Fist` and `Last` cases here instead (with a comment explaining why)?

GCC (and probably clang) complains about duplicate labels in the switch statement...

File: src/type-checker.cc (hunk 1 above)
Comment (keithw) [reply src/type-checker.cc:175]: OK, let me revisit this and see how that ends up looking.

File: src/wat-writer.cc (hunk 2 above)
Comment (sbc100) [reply src/wat-writer.cc]: If GCC can't give the error/warning fidelity we want here perhaps we should simply disable that particular warning under GCC?

File: src/type-checker.cc (hunk 1 above)
Comment (keithw) [reply src/type-checker.cc]: Here's an attempt to fix that callsite instead (which now adds some additional logged error messages to two tests).

File: src/wat-writer.cc (hunk 2 above)
Comment (keithw) [reply src/wat-writer.cc]: In the latest version, I kept the switches as-is and default-initialized the variable instead (in two places). It seems like it would be nice not to have to suppress "may be used uninitialized" warnings completely...?

File: test/typecheck/bad-callref-int32.txt [hunk 3]
```diff
@@ -12,4 +12,7 @@
 out/test/typecheck/bad-callref-int32.txt:6:6: error: type mismatch in call_ref, expected reference but got [i32]
     (call_ref (i32.const 10)
      ^^^^^^^^
+out/test/typecheck/bad-callref-int32.txt:6:6: error: type mismatch at end of function, expected [] but got [i32]
+    (call_ref (i32.const 10)
+     ^^^^^^^^
```
Comment (sbc100) [inline test/typecheck/bad-callref-int32.txt:17]: I'm no quite sure I understand why we now see more errors.  Is this a good thing?

**Review by sbc100 — APPROVED** [review]
lgtm % comment

File: test/typecheck/bad-callref-int32.txt (hunk 3 above)
Comment (keithw) [reply test/typecheck/bad-callref-int32.txt]: I don't think I understand the function-references implementation well enough to give a really considered answer, but I ended up finding a more minimal diff that also matches the current error messages 100%.
